File: find_duplicates.js
// find_duplicates.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { promisify } = require('util');
const glob = require('glob'); // npm install glob
const readdir = promisify(fs.readdir);
const stat = promisify(fs.stat);

const COLORS = {
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    reset: '\x1b[0m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

async function getFileHash(filepath, algo, partial) {
    return new Promise((resolve, reject) => {
        const hash = crypto.createHash(algo);
        const stream = fs.createReadStream(filepath);
        let bytesRead = 0;
        stream.on('data', (chunk) => {
            if (partial > 0 && bytesRead >= partial) {
                stream.destroy();
                resolve(hash.digest('hex'));
                return;
            }
            const toRead = partial > 0 ? Math.min(chunk.length, partial - bytesRead) : chunk.length;
            hash.update(chunk.slice(0, toRead));
            bytesRead += toRead;
        });
        stream.on('end', () => resolve(hash.digest('hex')));
        stream.on('error', reject);
    });
}

async function walk(dir, recursive, excludePatterns) {
    const results = [];
    const list = await readdir(dir);
    for (const item of list) {
        const full = path.join(dir, item);
        const st = await stat(full);
        if (st.isDirectory()) {
            // Исключения
            let skip = false;
            for (const pat of excludePatterns) {
                if (glob.sync(pat, { cwd: dir, strict: false }).includes(item)) {
                    skip = true;
                    break;
                }
            }
            if (skip) continue;
            if (recursive) {
                const sub = await walk(full, recursive, excludePatterns);
                results.push(...sub);
            }
        } else {
            // Исключение файлов
            let skip = false;
            for (const pat of excludePatterns) {
                if (glob.sync(pat, { cwd: dir, strict: false }).includes(item)) {
                    skip = true;
                    break;
                }
            }
            if (skip) continue;
            if (st.size > 0) {
                results.push({ path: full, size: st.size });
            }
        }
    }
    return results;
}

async function findDuplicates(root, recursive, algo, partial, excludePatterns) {
    const files = await walk(root, recursive, excludePatterns);
    const sizeMap = {};
    for (const f of files) {
        if (!sizeMap[f.size]) sizeMap[f.size] = [];
        sizeMap[f.size].push(f.path);
    }
    const duplicates = {};
    for (const size in sizeMap) {
        const group = sizeMap[size];
        if (group.length < 2) continue;
        const hashMap = {};
        for (const file of group) {
            const h = await getFileHash(file, algo, partial);
            if (!hashMap[h]) hashMap[h] = [];
            hashMap[h].push(file);
        }
        for (const h in hashMap) {
            if (hashMap[h].length > 1) {
                if (!duplicates[h]) duplicates[h] = [];
                duplicates[h].push({ size: parseInt(size), files: hashMap[h] });
            }
        }
    }
    return duplicates;
}

function formatSize(size) {
    const units = ['B', 'KB', 'MB', 'GB', 'TB'];
    let s = size;
    for (const u of units) {
        if (s < 1024) return `${s.toFixed(1)} ${u}`;
        s /= 1024;
    }
    return `${s.toFixed(1)} PB`;
}

function printDuplicates(duplicates, verbose) {
    const entries = Object.values(duplicates).flat();
    if (entries.length === 0) {
        console.log(colorize('Дубликатов не найдено.', 'green'));
        return;
    }
    let totalGroups = 0, totalFiles = 0, totalSize = 0;
    for (const entry of entries) {
        const { size, files } = entry;
        totalGroups++;
        totalFiles += files.length;
        totalSize += size * (files.length - 1);
        console.log(colorize(`\nГруппа дубликатов (размер: ${formatSize(size)}):`, 'blue'));
        console.log(`  Хеш: ${Object.keys(duplicates).find(k => duplicates[k].includes(entry))}`);
        for (let i = 0; i < files.length; i++) {
            if (i === 0) console.log(colorize(`  [оригинал] ${files[i]}`, 'green'));
            else console.log(colorize(`  [дубликат] ${files[i]}`, 'yellow'));
        }
        if (verbose) console.log(`  Количество файлов: ${files.length}`);
    }
    console.log(colorize(`\nВсего групп: ${totalGroups}, файлов: ${totalFiles}, экономия: ${formatSize(totalSize)}`, 'green'));
}

async function main() {
    const args = require('minimist')(process.argv.slice(2), {
        string: ['p', 'a', 'e', 'move', 'l'],
        boolean: ['r', 'delete', 'hardlink', 'n', 'v', 'y'],
        alias: { p: 'path', r: 'recursive', a: 'algo', e: 'exclude', n: 'dry-run', l: 'log', v: 'verbose' },
        default: { p: '.', r: true, a: 'md5', partial: 0 }
    });
    const root = path.resolve(args.p);
    const recursive = args.r;
    const algo = args.a;
    const partial = args.partial || 0;
    const excludePatterns = args.e ? args.e.split(',').map(s => s.trim()) : [];
    const deleteDups = args.delete;
    const moveTo = args.move;
    const hardlink = args.hardlink;
    const dryRun = args.n;
    const logFile = args.l;
    const verbose = args.v;
    const yes = args.y;

    try {
        const statInfo = await stat(root);
        if (!statInfo.isDirectory()) {
            console.log(colorize(`Ошибка: '${root}' не является папкой`, 'red'));
            process.exit(1);
        }
    } catch (err) {
        console.log(colorize(`Ошибка: ${err.message}`, 'red'));
        process.exit(1);
    }

    const dups = await findDuplicates(root, recursive, algo, partial, excludePatterns);
    printDuplicates(dups, verbose);

    if (deleteDups || moveTo || hardlink) {
        if (!yes) {
            const readline = require('readline').createInterface({
                input: process.stdin,
                output: process.stdout
            });
            const answer = await new Promise(resolve => {
                readline.question(colorize('Выполнить действия над дубликатами? [y/N] ', 'yellow'), resolve);
            });
            readline.close();
            if (answer.toLowerCase() !== 'y') {
                console.log(colorize('Операция отменена.', 'red'));
                return;
            }
        }
        for (const groups of Object.values(dups)) {
            for (const group of groups) {
                const original = group.files[0];
                for (let i = 1; i < group.files.length; i++) {
                    const dup = group.files[i];
                    if (dryRun) {
                        console.log(colorize(`[DRY RUN] Будет обработан дубликат: ${dup}`, 'yellow'));
                        continue;
                    }
                    if (deleteDups) {
                        fs.unlinkSync(dup);
                        console.log(colorize(`Удалён: ${dup}`, 'red'));
                    } else if (moveTo) {
                        fs.mkdirSync(moveTo, { recursive: true });
                        const dest = path.join(moveTo, path.basename(dup));
                        fs.renameSync(dup, dest);
                        console.log(colorize(`Перемещён: ${dup} -> ${dest}`, 'yellow'));
                    } else if (hardlink) {
                        fs.unlinkSync(dup);
                        fs.linkSync(original, dup);
                        console.log(colorize(`Создана жёсткая ссылка: ${dup} -> ${original}`, 'blue'));
                    }
                }
            }
        }
    }
}

main().catch(err => {
    console.log(colorize(`Ошибка: ${err.message}`, 'red'));
    process.exit(1);
});

File: find_duplicates.go
// find_duplicates.go
package main

import (
	"crypto/md5"
	"crypto/sha1"
	"flag"
	"fmt"
	"hash"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

const (
	reset  = "\033[0m"
	green  = "\033[92m"
	red    = "\033[91m"
	yellow = "\033[93m"
	blue   = "\033[94m"
)

func colorize(text, color string) string {
	return color + text + reset
}

type FileInfo struct {
	Path string
	Size int64
	Hash string
}

func getFileHash(path string, algo string, partial int64) (string, error) {
	var h hash.Hash
	switch algo {
	case "md5":
		h = md5.New()
	case "sha1":
		h = sha1.New()
	default:
		return "", fmt.Errorf("unsupported algorithm")
	}
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()
	if partial > 0 {
		// читаем только partial байт
		buf := make([]byte, partial)
		n, err := f.Read(buf)
		if err != nil && err != io.EOF {
			return "", err
		}
		h.Write(buf[:n])
	} else {
		_, err = io.Copy(h, f)
		if err != nil {
			return "", err
		}
	}
	return fmt.Sprintf("%x", h.Sum(nil)), nil
}

func findDuplicates(root string, recursive bool, algo string, partial int64, excludePatterns []string) (map[string][][]FileInfo, error) {
	sizeMap := make(map[int64][]string)
	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil // пропускаем ошибки доступа
		}
		if info.IsDir() {
			if !recursive && path != root {
				return filepath.SkipDir
			}
			// Исключение папок
			for _, pat := range excludePatterns {
				matched, _ := filepath.Match(pat, info.Name())
				if matched {
					return filepath.SkipDir
				}
			}
			return nil
		}
		// Исключение файлов
		for _, pat := range excludePatterns {
			matched, _ := filepath.Match(pat, info.Name())
			if matched {
				return nil
			}
		}
		if info.Size() > 0 {
			sizeMap[info.Size()] = append(sizeMap[info.Size()], path)
		}
		return nil
	})
	if err != nil {
		return nil, err
	}

	duplicates := make(map[string][][]FileInfo)
	for size, files := range sizeMap {
		if len(files) < 2 {
			continue
		}
		hashMap := make(map[string][]string)
		for _, f := range files {
			h, err := getFileHash(f, algo, partial)
			if err != nil {
				continue
			}
			hashMap[h] = append(hashMap[h], f)
		}
		for h, group := range hashMap {
			if len(group) > 1 {
				var infos []FileInfo
				for _, p := range group {
					info, _ := os.Stat(p)
					infos = append(infos, FileInfo{Path: p, Size: size, Hash: h})
				}
				duplicates[h] = append(duplicates[h], infos)
			}
		}
	}
	return duplicates, nil
}

func formatSize(size int64) string {
	units := []string{"B", "KB", "MB", "GB", "TB"}
	s := float64(size)
	for _, u := range units {
		if s < 1024.0 {
			return fmt.Sprintf("%.1f %s", s, u)
		}
		s /= 1024.0
	}
	return fmt.Sprintf("%.1f PB", s)
}

func printDuplicates(duplicates map[string][][]FileInfo, verbose bool) {
	if len(duplicates) == 0 {
		fmt.Println(colorize("Дубликатов не найдено.", green))
		return
	}
	var totalGroups, totalFiles int
	var totalSize int64
	for _, groups := range duplicates {
		for _, group := range groups {
			totalGroups++
			totalFiles += len(group)
			size := group[0].Size
			totalSize += size * int64(len(group)-1)
			fmt.Println(colorize(fmt.Sprintf("\nГруппа дубликатов (размер: %s):", formatSize(size)), blue))
			fmt.Printf("  Хеш: %s\n", group[0].Hash)
			for i, fi := range group {
				if i == 0 {
					fmt.Println(colorize(fmt.Sprintf("  [оригинал] %s", fi.Path), green))
				} else {
					fmt.Println(colorize(fmt.Sprintf("  [дубликат] %s", fi.Path), yellow))
				}
			}
			if verbose {
				fmt.Printf("  Количество файлов: %d\n", len(group))
			}
		}
	}
	fmt.Println(colorize(fmt.Sprintf("\nВсего групп: %d, файлов: %d, экономия: %s", totalGroups, totalFiles, formatSize(totalSize)), green))
}

func main() {
	var (
		path    string
		recursive bool
		algo    string
		partial int64
		exclude string
		deleteDups bool
		moveTo  string
		hardlink bool
		dryRun  bool
		logFile string
		verbose bool
		yes     bool
	)
	flag.StringVar(&path, "p", ".", "Папка для сканирования")
	flag.BoolVar(&recursive, "r", true, "Рекурсивно")
	flag.StringVar(&algo, "a", "md5", "Алгоритм: md5 или sha1")
	flag.Int64Var(&partial, "partial", 0, "Хешировать только первые N байт")
	flag.StringVar(&exclude, "e", "", "Исключить по glob-шаблону (через запятую)")
	flag.BoolVar(&deleteDups, "delete", false, "Удалить дубликаты")
	flag.StringVar(&moveTo, "move", "", "Переместить дубликаты в папку")
	flag.BoolVar(&hardlink, "hardlink", false, "Заменить дубликаты жёсткими ссылками")
	flag.BoolVar(&dryRun, "n", false, "Симуляция")
	flag.StringVar(&logFile, "l", "", "Лог-файл")
	flag.BoolVar(&verbose, "v", false, "Подробно")
	flag.BoolVar(&yes, "y", false, "Не запрашивать подтверждение")
	flag.Parse()
	if flag.NArg() > 0 {
		path = flag.Arg(0)
	}

	var excludePatterns []string
	if exclude != "" {
		excludePatterns = strings.Split(exclude, ",")
	}

	root, err := filepath.Abs(path)
	if err != nil {
		fmt.Println(colorize("Ошибка: "+err.Error(), red))
		os.Exit(1)
	}
	info, err := os.Stat(root)
	if err != nil || !info.IsDir() {
		fmt.Println(colorize("Ошибка: '"+root+"' не является папкой", red))
		os.Exit(1)
	}

	dups, err := findDuplicates(root, recursive, algo, partial, excludePatterns)
	if err != nil {
		fmt.Println(colorize("Ошибка сканирования: "+err.Error(), red))
		os.Exit(1)
	}
	printDuplicates(dups, verbose)

	// Действия
	if deleteDups || moveTo != "" || hardlink {
		if !yes {
			fmt.Print(colorize("Выполнить действия над дубликатами? [y/N] ", yellow))
			var ans string
			fmt.Scanln(&ans)
			if strings.ToLower(ans) != "y" {
				fmt.Println(colorize("Операция отменена.", red))
				return
			}
		}
		for _, groups := range dups {
			for _, group := range groups {
				original := group[0]
				for _, dup := range group[1:] {
					if dryRun {
						fmt.Println(colorize(fmt.Sprintf("[DRY RUN] Будет обработан дубликат: %s", dup.Path), yellow))
						continue
					}
					if deleteDups {
						os.Remove(dup.Path)
						fmt.Println(colorize(fmt.Sprintf("Удалён: %s", dup.Path), red))
					} else if moveTo != "" {
						os.MkdirAll(moveTo, 0755)
						dest := filepath.Join(moveTo, filepath.Base(dup.Path))
						os.Rename(dup.Path, dest)
						fmt.Println(colorize(fmt.Sprintf("Перемещён: %s -> %s", dup.Path, dest), yellow))
					} else if hardlink {
						os.Remove(dup.Path)
						os.Link(original.Path, dup.Path)
						fmt.Println(colorize(fmt.Sprintf("Создана жёсткая ссылка: %s -> %s", dup.Path, original.Path), blue))
					}
				}
			}
		}
	}
}
